### ISSUE TITLE
Add different proxy render presets to build_ffmpeg_proxies

### ISSUE THREAD

**Issue by NathanLovato** [GDquest/Blender-power-sequencer#174]
If you have experience with video encoding, we need your help here! I'd like to offer some solid presets to generate different types of proxy files for blender. They should let the user favor a small size, high encoding speed, high decoding speed in Blender and high color fidelity, depending on their needs.

The current preset favors a small size and high encoding speed as I've mostly edited screencasts/tutorials so far. But the colors aren't too great and the picture is a little blurry. Blender's default JPEG-based proxies aren't too great either imo.

I've got a contact from the CG industry who worked with video technologies for decades. If no one can help here, I'll ask him for some insights.

- [ ] Allow the user to render 100, 50, and 25% proxies
- [ ] h264 proxies for small size
- [ ] MJPEG or something for color fidelity? IIRC prores was a little too heavy for Blender

The script could use some refactoring, getting broken down as well... waiting for razvan to run me through his work on the multithreaded render cli tool.

**Comment by jooert** [comment]
I'm definitely no expert but I did some "research" on the internet a while ago. What I gathered is that you want to use an intra-frame codec (like MJPEG, ProRes, DNxHD) instead of an inter-frame codec (MPEG2, H264); [currently the proxy script uses MPEG2](https://github.com/GDquest/Blender-power-sequencer/blob/master/scripts/build_ffmpeg_proxies.py#L66). The difference is that with an intra-frame codec, only the information in a single frame is used to compress it. That makes scrubbing through footage really smooth as to render a frame no other frame than the current one has to be fetched. The downside is that files are bigger as every frame is stored on its own even if nothing has changed in between two frames.

I settled for DNxHD (can't remember exactly why; I think it offered better quality than MJPEG and ProRes was a little wonky) and use a command like
```shell
ffmpeg -i <input_file> -c:v dnxhd -b:v 36M <output_name>.mxf
```
to transcode my footage. Only certain bitrates are supported, see [this Wikipedia page](https://en.wikipedia.org/wiki/List_of_Avid_DNxHD_resolutions) for the allowed ones.

Even though I use the original resolution (100%) this has a huge impact on playback performance in Blender. On my ancient 2008 Core 2 Duo, the framerate improves from 5-6 fps with the original FullHD 50Mbps H264 file to 23-25 fps with the 36Mbps DNxHD one. Scrubbing is practically impossible with the original footage whereas it is smooth as butter with the transcoded files.

I hope this might be useful to you.

**Comment by NathanLovato** [comment]
Thanks much! Yup I'm aware of that. inter-frame codecs can work like intra-frame ones, you can control the number of frames you encode together (gopsize in ffmpeg). The current preset is designed for tablets/laptops in mind: I work with a 500gd drive, so 1gb/minute is too heavy. And in my experience working with USB3, the connection is a bottleneck.

I wouldn't use dnxhd because it only supports 1080p and 720p. The two options left for high quality  are Prores from apple, they have a special codec for video proxies - although I haven't check if ffmpeg supports it. But it should support any resolution. And there's MJPEG. And IIRC for audio, PCM is the fastest format.

libx264 proxies may also be nice to have as an option for 25%/50%, for optimal size and fastest generation speed. But it needs some tests for playback accuracy - you may have a few frames offset depending on the project

**Comment by jooert** [comment]
I hope this didn't come across as me trying to lecture you on something you already knew! What I was trying to get at is that intra-frame codecs (in my experience) improve the performance more than just reducing the resolution. So switching to such a codec (or reducing the GOP size) would be the first thing I would try for improved proxies.

Only 1080p/720p in DNxHD is definitely problematic for a general-purpose proxy solution. MJPEG is basically just JPEG in a video container so I'm not sure if the quality will be much better than the Blender-generated proxies (although one can [tweak the quality setting when encoding with ffmpeg](https://stackoverflow.com/questions/32147805/ffmpeg-generate-higher-quality-images-for-mjpeg-encoding#32151594)). There seems to be support for the [ProRes proxy profile in ffmpeg](https://ffmpeg.org/ffmpeg-codecs.html#ProRes) but I haven't tried it yet.

For audio, I would have expected PCM to be the fastest as it is just uncompressed audio. Although of course larger in file size than other formats this shouldn't be a problem in comparison to the situation with video – a mono track of 16-bit PCM audio is just 5 MB per minute.

> libx264 proxies may also be nice to have as an option for 25%/50%, for optimal size and fastest generation speed. But it needs some tests for playback accuracy - you may have a few frames offset depending on the project.
That's something I don't understand, could you explain why/when this will happen?

**Comment by jooert** [comment]
I just did a quick test comparing MJPEG and ProRes Proxy. Encoding was significantly faster for MJPEG (1.5x vs 0.6x). Futhermore, MJPEG produced – at least for the test footage I was using –  proxies of better quality (less blurry) when aiming for comparable file size. (ProRes does support 10-bit video though.)

**Comment by NathanLovato** [comment]
François got back to me and shared some more insights on proxies. He recommends to play with mp4/h264 and webm instead of Mjpeg for proxies (NB: he's worked with video encoding and for TV for a long time). This way we can play with simple parameters, like CRF and a low gopsize. I'll start experimenting later today

**Comment by Francois-Grassard** [comment]
Hi everybody,


As Nathan said, I work for broadcast industry since 20 years now and I use FFMPEG everyday for so many tasks, including encoding/transcoding, essences extraction, rewraping of multiple audio tracks, colorspaces conversion, etc …

I hope I can help with my knowledge for the topic of proxies generation. :)

Because proxies have to be small, you can definitivly forget ProRes and DnxHD. Even if the last one is open-sourced, it is not designed to be used for that. The lowest profile available for this codec is still using 36 Mbps ! That’s way too much for this purpose. DnxHD is also only available for the MXF and Quicktime containers. The first one is not well known by the public and the second one is the worst container we could used today, for so many reasons (color shifting, gamma issues, metadatas about the footage saved at the end of the file instead of the header, etc …).
  
The main purpose of ProRes and DnxHD is to be able to reuse initial macroblocks information of the source media, to avoid loosing too much information when you have to re-encode the same image multiple times. The counterpart is the big file size, even when you select the lowest profile.

MJPEG has the advantage to be fast to decode. But it only use an old static DCT algorithm for each image and don’t optimize file size using temporal redundancy. It was the best solution 10 years ago, but not today when you want to find the best balance between quality, reading efficiency and size of the files.


So, I think we have to split the problem in too parts. The container and the codec.

For the container, my recommandation is to provide this two solutions :

- **WebM** : This container is open/free/libre, wich is a good point if you only want to use free formats. It is also the container of the future, who will be used with the new open codec AV1, alongside with « Opus » for the audio. In 2020/2022, I bet on the progressive fall of H.264/H.265 contained in MP4, and a raise of the AV1 contained in WebM. I’ll read this post in 3 years to see if I was right. :)

- **MP4** : Officialy closed but used by a lot of people, MP4 can also be read by Mac users without installing any other application (such as VLC). AFAIK, that’s not the case for WebM. To be clear, I only talk about the container and NOT the MP4 codec, here (who exists in different versions, such as the h.264).


For the Codec, my recommandation is to provide this two solutions :

- **VP8** : This codec has been developped alongside the WebM Container and it’s also open/free/libre. The last codec of this familly is the VP9, but it is very very slow compared to VP8. For the purpose of proxies, I prefer to ignore it.

- **h.264** : This codec is today the most known for video encoding. Even if it’s not open/free/libre, I think it is a good idea to propose it, for the same reasons we will provide MP4 as a container.


So, if you wan’t to only use open formats, you will go with WebM/VP8 and if you don’t, you can go with MP4/h.264.

The most important FFMPEG options for this two formats are the followings :

Common parameters for h.264 and VP8 :

- **Pixel format** : Describe the number of planes/channels used in the video and the chroma-subsampling scheme. Because Quicktime player and Windows Media Player can’t read 444 pixel format or full RGB, is it a good idea to restrict the video to 420. So, we gonna used -pix_fmt yuv420p
- 
- **Color Matrix** : If you want to avoid color shifting during the encoding of the proxies, you have to specify the target colorspace. The most common colorspace used today in video is bt601(SD) / bt709 (HD). So we gonna force a conversion to this specific colorspace by using -vf colormatrix=bt601:bt709
- 
- **GOP Size** :  To be short and not too technical, VP8 and h.264 group image in GOPs (Group Of Picture). The longer they are, the slower will your computer be to decode them and the lower the file size will be also (that’s the only good point). If they are shorts, your computer will requiert less processing to decode them and the playing of the proxies will not be laggy. I recommand a GOP Size of 1 or 2 in the specific case of Blender (prefer 1 to be fast in any case), and never more than 6 in general. The syntax to define GOP 1 is -g 1
- 
- **Constant Rate Factor** : Control the biterate of a video stream. You should forget the old CBR and VBR scheme and use CRF. To be short, CRF is like a « Quality factor », define between 0 and 51. When you are close at 0, you are really close to Lossless (h.264 has a real Lossless profile, that’s why I use the « close » term). When you reach 51, your video is really degrated … you will probably don’t recognize anyone in the image. I usually choose a value between 20 and 30 and recommand you use 25. The quality will be OK for most of the contents.


Specific parameters to h.264 :

- One of the good part of FFMPEG when you want to encode a MP4/h.264, is that you don’t have to specify the codec. You just have to used the *.mp4 as extension and FFMPEG will automaticaly choose the x.264 encoder (free implementation of the h.264 standard). Good point !
- 
- **Preset** : To be short again, if you let enough time to the encoder to find a good motion vectors optimization in the the video, the file will be smaller. If you hurry it (that could be the case for proxies generation), the quality will be the same, but the file will be bigger. Between the two, you have access to different flavours : ultrafast / superfast / veryfast / faster / fast / medium / slow / slower / veryslow. By default, if you don’t specify a preset, FFMPEG use the medium one. FYI, the difference between ultrafast and veryslow is HUGE. The file could be 50 times bigger from one preset to another !



Specific parameters to VP8 :

- **Codec** : You have to specify wich codec to use in WebM, bewteen VP8 and VP9. For the first one, just type the name of the dedicated encoder with -c:v libvpx (just to let you know, is you want to use VP9, you have to type « libvpx-vp9 ».)
- 
- **Threads** : By default, h.264 use all cores available for the encoding process, but VP8 only use one core in FFMPEG. If you wan’t to use more than one, you have to use for example -threads 8 (if you have 8 cores of course … maybe it could be handy to add this option in the interface to set this parameter, if you can detect them automaticaly).
- 
- **Speed** : Similar to Preset in h.264. Define how long will it take to optimize the motion vectors. The only difference is that the quality of the image can change a bit between values. The range is -16 to 16. 16 is the fastest one (similar to Ultrafast in h.264).


So, here is a typical command-line for each Container/Codec :

WebM / VP8 :

ffmpeg -i input.mov -pix_fmt yuv420p -crf 25 -c:v libvpx -speed 16 -g 1 -threads 8 -vf colormatrix=bt601:bt709 output.webm

MP4 / h.264 :

ffmpeg -i input.mov -pix_fmt yuv420p -crf 25 -preset veryslow -g 1 -vf colormatrix=bt601:bt709 output.mp4

Now, you know exactly what is the task of each flag. You can modulated what you want and provide multiple presets such as :

- Fast encoding / Bigger files / Low Quality / Short GOP
- Slow encoding / Smaller files / Better Quality / Mediaum GOP
- Etc …

I Hope this infomation will help you. Cheers !

**Comment by 3pointedit** [comment]
Well even if I never use the proxies this email was exceptionally helpful!
That was really informative Francois, and having access to quality values
would be helpful. I wonder if proxy generation could occur as a background
operation or a downtime function? Can we use proxies to prerender effects
so they play realtime (like a buffer), is proxy on effect strip possible? I
don't recall 🙄

**Comment by NathanLovato** [comment]
Closed with 5559e0d5814ea4922c0f6b1b057ac2374643eaec and ac8533e2a0b966bbf2e3cb15b381d2838e884e60
